core/src/ledger_cleanup_service.rs

```diff
@@ -62,37 +65,88 @@ impl LedgerCleanupService {
-
-        if let (Ok(disk_utilization_pre), Ok(disk_utilization_post)) =
-            (disk_utilization_pre, disk_utilization_post)
-        {
-            datapoint_debug!(
-                "ledger_disk_utilization",
-                ("disk_utilization_pre", disk_utilization_pre as i64, i64),
-                ("disk_utilization_post", disk_utilization_post as i64, i64),
-                (
-                    "disk_utilization_delta",
-                    (disk_utilization_pre as i64 - disk_utilization_post as i64),
-                    i64
-                )
+            // Notify blockstore of impending purge
+            //cleanup
+            let mut shreds = Vec::new();
+            let mut iterate_time = Measure::start("iterate_time");
+            let mut total_shreds = 0;
+            for (i, (slot, meta)) in blockstore.slot_meta_iterator(0).unwrap().enumerate() {
+                if i == 0 {
+                    debug!("purge: searching from slot: {}", slot);
+                }
+                total_shreds += meta.completed_data_indexes.len();
+                shreds.push((slot, meta.completed_data_indexes.len()));
+                if slot > root {
+                    break;
+                }
+            }
+            iterate_time.stop();
+            debug!(
+                "checking for ledger purge: {:?} max: {} slots: {} total_shreds: {} {}",
+                disk_utilization_pre,
+                max_ledger_utilization,
+                shreds.len(),
+                total_shreds,
+                iterate_time
             );
+            if (total_shreds as u64) < max_ledger_utilization {
+                return Ok(());
+            }
+            let mut cur_shreds = 0;
+            let mut lowest_slot_to_clean = shreds[0].0;
+            for (slot, num_shreds) in shreds.iter().rev() {
+                cur_shreds += *num_shreds as u64;
+                if cur_shreds > max_ledger_utilization {
+                    lowest_slot_to_clean = *slot;
+                    break;
+                }
+            }
+            if cur_shreds > 0 {
+                debug!(
+                    "cleaning up to: {} shreds: {}",
+                    lowest_slot_to_clean, total_shreds
+                );
+                let mut slot_update_time = Measure::start("slot_update");
+                *blockstore.lowest_cleanup_slot.write().unwrap() = lowest_slot_to_clean;
+                slot_update_time.stop();
+                let mut clean_time = Measure::start("ledger_clean");
+                blockstore.purge_slots(0, Some(lowest_slot_to_clean));
+                clean_time.stop();
+                debug!("ledger purge: {} {}", slot_update_time, clean_time);
+            }
+
+            let disk_utilization_post = blockstore.storage_size();
+
+            if let (Ok(disk_utilization_pre), Ok(disk_utilization_post)) =
+                (disk_utilization_pre, disk_utilization_post)
+            {
+                datapoint_debug!(
```

Comment:
do `debug` level datapoints actually get logged now? For a while it was only info and above

Reply:
`datapoint_debug` don't get logged by default